File: hello-grpc-flutter/lib/conn/web_grpc_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:fixnum/fixnum.dart';
import '../common/landing.pb.dart';
import '../common/common.dart';

class WebGrpcClient {
  final String baseUrl;
  
  WebGrpcClient(String host, int port) : baseUrl = 'http://$host:9997'; // 连接到HTTP网关
  
  Future<TalkResponse> talk(TalkRequest request) async {
    final response = await http.post(
      Uri.parse('$baseUrl/api/talk'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'data': request.data,
        'meta': request.meta,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return _buildTalkResponseFromJson(data);
    } else {
      throw Exception('HTTP ${response.statusCode}: ${response.body}');
    }
  }
  
  Stream<TalkResponse> talkOneAnswerMore(TalkRequest request) async* {
    final response = await http.post(
      Uri.parse('$baseUrl/api/talkOneAnswerMore'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'data': request.data,
        'meta': request.meta,
      }),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final List<dynamic> results = data['results'] ?? [];
      for (final result in results) {
        yield _buildTalkResponseFromJson(result);
      }
    } else {
      throw Exception('HTTP ${response.statusCode}: ${response.body}');
    }
  }
  
  Future<TalkResponse> talkMoreAnswerOne(Stream<TalkRequest> requests) async {
    final List<Map<String, dynamic>> requestList = [];
    await for (final request in requests) {
      requestList.add({
        'data': request.data,
        'meta': request.meta,
      });
    }
    
    final response = await http.post(
      Uri.parse('$baseUrl/api/talkMoreAnswerOne'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'requests': requestList}),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      return _buildTalkResponseFromJson(data);
    } else {
      throw Exception('HTTP ${response.statusCode}: ${response.body}');
    }
  }
  
  Stream<TalkResponse> talkBidirectional(Stream<TalkRequest> requests) async* {
    final List<Map<String, dynamic>> requestList = [];
    await for (final request in requests) {
      requestList.add({
        'data': request.data,
        'meta': request.meta,
      });
    }
    
    final response = await http.post(
      Uri.parse('$baseUrl/api/talkBidirectional'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'requests': requestList}),
    );
    
    if (response.statusCode == 200) {
      final data = jsonDecode(response.body);
      final List<dynamic> results = data['results'] ?? [];
      for (final result in results) {
        yield _buildTalkResponseFromJson(result);
      }
    } else {
      throw Exception('HTTP ${response.statusCode}: ${response.body}');
    }
  }
  
  TalkResponse _buildTalkResponseFromJson(Map<String, dynamic> data) {
    final response = TalkResponse()
      ..status = data['status'] ?? 200;
    
    final List<dynamic> results = data['results'] ?? [];
    for (final resultData in results) {
      final result = TalkResult()
        ..id = Int64.parseInt(resultData['id']?.toString() ?? '0')
        ..type = _parseResultType(resultData['type']);
      
      // 处理kv映射
      final Map<String, dynamic> kv = resultData['kv'] ?? {};
      kv.forEach((key, value) {
        result.kv[key] = value.toString();
      });
      
      response.results.add(result);
    }
    
    return response;
  }
  
  ResultType _parseResultType(dynamic type) {
    if (type == null) return ResultType.OK;
    if (type is String) {
      switch (type.toUpperCase()) {
        case 'OK':
          return ResultType.OK;
        case 'FAIL':
          return ResultType.FAIL;
        default:
          return ResultType.OK;
      }
    }
    if (type is int) {
      return type == 0 ? ResultType.OK : ResultType.FAIL;
    }
    return ResultType.OK;
  }
}
}

File: hello-grpc-tauri/src/modules/web_grpc_client.js
// Web gRPC Client for HTTP Gateway communication
// Similar to Flutter's WebGrpcClient

class WebGrpcClient {
    constructor(host, port) {
        this.baseUrl = `http://${host}:9997`; // Connect to HTTP gateway
    }

    async talk(request) {
        const response = await fetch(`${this.baseUrl}/api/talk`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                data: request.data,
                meta: request.meta
            })
        });

        if (response.ok) {
            const data = await response.json();
            return this._buildTalkResponseFromJson(data);
        } else {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }
    }

    async* talkOneAnswerMore(request) {
        const response = await fetch(`${this.baseUrl}/api/talkOneAnswerMore`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                data: request.data,
                meta: request.meta
            })
        });

        if (response.ok) {
            const data = await response.json();
            const results = data.results || [];
            for (const result of results) {
                yield this._buildTalkResponseFromJson(result);
            }
        } else {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }
    }

    async talkMoreAnswerOne(requests) {
        const requestList = [];
        for await (const request of requests) {
            requestList.push({
                data: request.data,
                meta: request.meta
            });
        }

        const response = await fetch(`${this.baseUrl}/api/talkMoreAnswerOne`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ requests: requestList })
        });

        if (response.ok) {
            const data = await response.json();
            return this._buildTalkResponseFromJson(data);
        } else {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }
    }

    async* talkBidirectional(requests) {
        const requestList = [];
        for await (const request of requests) {
            requestList.push({
                data: request.data,
                meta: request.meta
            });
        }

        const response = await fetch(`${this.baseUrl}/api/talkBidirectional`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ requests: requestList })
        });

        if (response.ok) {
            const data = await response.json();
            const results = data.results || [];
            for (const result of results) {
                yield this._buildTalkResponseFromJson(result);
            }
        } else {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }
    }

    _buildTalkResponseFromJson(data) {
        // Create a response object that matches the expected format
        return {
            id: data.id,
            status: data.status || 200,
            results: data.results || data.kv || data
        };
    }
}

// Export for use in main.js
window.WebGrpcClient = WebGrpcClient;
